[MLIR][Linalg] Remove matmul_transpose variants

rengolin opened the pull request:
Removes the `(batch_)matmul_transpose_{a|b}` variants from OpDSL and replace it with `matmul affine_maps [...]` whenever appropriate. This is in line with the [plan](https://discourse.llvm.org/t/rfc-op-explosion-in-linalg/82863), and can be done since #104783 merged.

See: https://discourse.llvm.org/t/deprecate-batch-matmul-transpose-a-b-linalg-operations/87245

Issues investigated:
* pad transform tests that could use `matmul` instead, so change to that.
* ArmSME test using transpose actually needed it, so changed to `matmul` + affine maps.

Arm tests validated by @banach-space (thanks!!).

rengolin commented:
> Could we get some time to manage the churn with these changes?

Of course!! No rush from my side. Let me know if you need some help.

qedawkins commented on mlir/lib/Dialect/Linalg/Transforms/TransposeMatmul.cpp:
```diff
@@ -57,18 +58,31 @@ FailureOr<Operation *> mlir::linalg::transposeMatmul(RewriterBase &rewriter,
+    newRHS = transposeOp->getResult(0);
+    mapLHS = AffineMap::get(3, 0, {d0, d2}, context);
+    mapRHS = AffineMap::get(3, 0, {d1, d2}, context);
+    mapOut = AffineMap::get(3, 0, {d0, d1}, context);
```
Pattern matching can't be replaced as easily, but we can add bespoke C++ for it like `matchMatmulTransposeB`

qedawkins replied:
> builder.create<linalg::MatmulOp>(..., transpose_a | broadcast_b);

This sounds good too. I mainly want to minimize the number of places downstream users need to setup affine maps. A specialization would be completely no change, but enums would be pretty lightweight (I suspect every frontend will end up redesigning the same "map generator" util if we don't provide it).

rengolin replied:
Ok, turns out the specializations make the changes in `TransposeMatmul.cpp` moot, so I reverted and it's not back to the original file. This means anyone using the builders out there will continue to have the same behaviour. 

So we're down to two problems: previous IR that needs regex-changing, and the pattern matchers. The IR changes are specific to each downstream, so better done in there as a one off. The pattern matchers will also be specific to usage and additional attributes, so can be on the downstreams until we figure a generic way to hold them all.

rengolin commented:
> It would be ideal if we had a deprecation path instead of just a switch.

Ack. I tried to gather that in the forum, but people were not paying attention to that, either. :smiley:

Also, this has been discussed for years and the original patch was almost 1 year ago. 

> * Torch-MLIR (and all other "front end dialects" like StableHLO, etc.) will need to be updated to not generate the batch-matmul operations

This should be straightforward. This PR has the recipe: replace `*_matmul_transpose_*` with `matmul` + appropriate affine map. Everything else _just works_. I could help with that, too, if needed.

> * and downstream compilers (like IREE) need to make sure that they can navigate away from this.

This is less straightforward. From the discussion back then, I had assumed we would all work towards not using it internally, so that we could easily switch later. This is 1 year later. :frowning_face: 

> In LLVM this would be something that would be deprecated on one release and removed on the next release (it is pretty much a break the world change). I dont think all the front end dialect folks are paying attention to this.

Not exactly. We have [backwards compatibility for LLVM IR](https://llvm.org/docs/DeveloperPolicy.html#ir-backwards-compatibility), and the support policy describes how to [deprecate entire components](https://llvm.org/docs/SupportPolicy.html#steps-for-removal), but those points apply. We neither have backwards compatibility guarantees for MLIR, nor we're talking about removing an entire dialect. Changes in dialect operations occur more often, and to be honest, this one is already turning 1 year old. I have just removed some old element-wise operations and it didn't _break the world_.

I'm assuming IREE isn't really using releases, here, so talking about releases doesn't make a lot of sense. But that doesn't mean we can be careless when removing stuff, it just means we can do _best effort_ on a case by case basis. We also need to balance upstream versus downstream needs, where the latter cannot impose hard constraints on the former, but can suggest time-frames and ways around.

My proposal is to fix `torch-mlir` to stop emitting `matmul_transpose_*` and understand what are the implications inside IREE (and any other linalg user that still relies on those operations). If IREE can be fixed quickly, we wait. If not, IREE can duplicate those ops inside `linalgx` and deprecate on your own speed.

rengolin commented:
> * Torch-MLIR (and all other "front end dialects" like StableHLO, etc.) will need to be updated to not generate the batch-matmul operations

I have just noticed you're talking here about the `match_matmul` variants, which are not the subject of this PR. Here I'm only removing the transposed variants, which do not seem to be anywhere in `torch-mlir`. Does that mean this PR is _"fine"_ from your point of view, or do you still have concerns specifically with the transposed variants?

rengolin commented:
I see a lot of usage of matmul transpose in IREE: https://github.com/search?q=repo%3Airee-org%2Firee%20matmul_transpose&type=code

A few things I can see from just that search:
* Some usages are "representational", ie. could easily be replaced with the alternative.
* `if (!isa<linalg::MatmulTransposeBOp, linalg::BatchMatmulTransposeBOp>(op)) {` could be replaced with the matcher that @qed discusses above. Here, just check something like `isMapTranspose` that works on both `matmul` and `batch_matmul`.
* I assume `{lowering_config = ...}` attribute could also be applied to `matmul`
* Upstream tests show that transforms/passes still work on the alternative, so IREE should benefit from those.
* IREE specific passes may need to adapt, but if made on par with upstream, may also work out of the box.

javedabsar1 commented on mlir/test/Dialect/Linalg/tile-to-forall.mlir:
```diff
@@ -504,7 +504,7 @@ func.func @matmul_tile_size_dynamic(%A: tensor<?x?xf32>, %B: tensor<?x?xf32>, %C
 
 module attributes {transform.with_named_sequence} {
   transform.named_sequence @__transform_main(%arg1: !transform.any_op {transform.readonly}) {
-    %0 = transform.structured.match ops{["linalg.matmul_transpose_b"]} in %arg1 : (!transform.any_op) -> !transform.any_op
```
Was the previous IR wrong? It seemed looking for matmul_transpose_b but line 500 was never that. I may be missing something.

rengolin replied:
Yup. The original commit (17 months ago!) had no transpose intent, so I'm guessing this was just a typo. And since this was expecting an error, there were no CHECK lines to break.

qedawkins commented:
> I see a lot of usage of matmul transpose in IREE: https://github.com/search?q=repo%3Airee-org%2Firee%20matmul_transpose&type=code
> 
> A few things I can infer from just that search:
> 
>     * Some usages are "representational", ie. could easily be replaced with the alternative.
> 
>     * `if (!isa<linalg::MatmulTransposeBOp, linalg::BatchMatmulTransposeBOp>(op)) {` could be replaced with the matcher that @qed discusses above. Here, just check something like `isMapTranspose` that works on both `matmul` and `batch_matmul`.
> 
>     * I assume `{lowering_config = ...}` attribute could also be applied to `matmul`
> 
>     * Upstream tests show that transforms/passes still work on the alternative, so IREE should benefit from those.
> 
>     * IREE specific passes may need to adapt, but if made on par with upstream, may also work out of the box.
> 
> 
> Of course, my assumptions may be wrong, this is just me being proactive. 😃

In theory everything there is coverage related or test related. Because Linalg is IREE's input, we have to support the existing named variants. We also try to retain the name for as long as possible for convenience and readability purposes, but also that gives ostensible reliance on those ops per your search.

What I believe Mahesh is primarily asking for is a grace period to align that illusion with reality. I see largely two paths:
1. Introduce the replacement (linalg.matmul w/ indexing maps exists and is that) and provide a deprecation period. That is unusual for MLIR though.
2. Design the C++ API changes to be as non-intrusive as possible. Things like bespoke matcher replacements and specialized builders make the majority of the changes purely mechanical and mitigates the surface area for bug finding along the way. This is mainly what I was asking for in the above comment.

Last comment I'll add about IREE, there is a fair amount of historical code that is currently unmaintained. Current development has been careful about parity between named and generic variants, but the old stuff wasn't necessarily. With that said, when old unused stuff breaks it gets deleted so not a huge deal :)

Honestly the part about this change I'm looking forward to the least is all the random IR snippets I have stashed away as a personal testing bank that I'll have to go update :P. Otherwise the direction here is great!

rengolin commented:
> 1. Introduce the replacement (linalg.matmul w/ indexing maps exists and is that) and provide a deprecation period. That is unusual for MLIR though.

Right, this is kinda what we did last year, minus the part where we agreed on a date. I'm trying to se the date now, but I don't think we should wait another year. :smiley: 

> 2. Design the C++ API changes to be as non-intrusive as possible. Things like bespoke matcher replacements and specialized builders make the majority of the changes purely mechanical and mitigates the surface area for bug finding along the way. This is mainly what I was asking for in the above comment.

I like this very much. But I also want to try to avoid too much _bespoke matchers_ because that explodes as well. I'm ok with specializations for the common cases, but that causes confusion for the non-common cases. I'm ok with C++ functions that match something, but then everyone adds theirs and we get duplicates. So, some plan here would be good to have.

> Current development has been careful about parity between named and generic variants, but the old stuff wasn't necessarily. With that said, when old unused stuff breaks it gets deleted so not a huge deal

I empathise. That's why I proposed a `linalgx.matmul_transpose_*` to be a stop-gap if you find code that is too hairy and can't be deleted straight away.

> Honestly the part about this change I'm looking forward to the least is all the random IR snippets I have stashed away as a personal testing bank that I'll have to go update

That's why we created the `mlir-gen` [tool](https://github.com/libxsmm/tpp-mlir/tree/main/tools/mlir-gen), but now I have no idea what half of the code is trying to do! :rofl: 

> Any way we can stage it so torch-mlir gets to go first would be very much appreciated for us. _I think_ it's just `linalg.matmul` in torch-mlir proper, but there are adjacent shark projects that embed hand written IR that use the transpose variants.

I see. How do we sync those changes with upstream?

banach-space reviewed (commented):
Thanks for doing this, it's long over-due and a much appreciate clean-up 🙏🏻  

The approach that you are taking here makes sense to me, though there should be some deprecation note for the specialised builders for `linalg.matmul` - presumably we don't want to keep them forever?

Btw, we rely on `linalg.matmul_transpose_a` for SME and as you mention in the summary, "upstream" LG. I also checked in IREE and identified a few minor things, but nothing too concerning. I will report in https://github.com/iree-org/iree/issues/21349.

kuhar commented on mlir/include/mlir/Dialect/Linalg/IR/Linalg.h:
```diff
@@ -144,4 +144,126 @@ std::pair<int64_t, int64_t> getFmrFromWinogradConv2DFmr(WinogradConv2DFmr fmr);
+namespace mlir::linalg {
+
+/// Specialization of `linalg.matmul` op that has a transpose map on A
+class MatmulTransposeAOp : public MatmulOp {
```
Should we immediately mark this as deprecated with `LLVM_DEPRECATED`? Same for the other transition classes below. IIUC, the intention is for this to be short-lived.

kuhar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3881,6 +3882,172 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+                MatmulOp::getRegionBuilder(), getAffineMaps(builder));
+}
+
+bool MatmulTransposeAOp::classof(Operation *op) {
+  return dyn_cast_or_null<linalg::MatmulOp>(op);
```
This seems like a footgun to me because it does not check for indexing maps. For example, if someone uses a builder for `MatmulTransposeA` and then do `isa<MatmulTransposeB>(trA)`, this will return `true` unexpectedly...

If we are not going to check the indexing maps, I think we should add a comment with a prominent warning at the very least.

kuhar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3881,6 +3882,172 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+
+SmallVector<AffineMap> MatmulTransposeBOp::getAffineMaps(OpBuilder &builder) {
+  AffineExpr d0, d1, d2;
+  auto context = builder.getContext();
```
nit
```suggestion
  MLIRContext *context = builder.getContext();
```
following https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable

kuhar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3881,6 +3882,172 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+  AffineMap mapLHS = AffineMap::get(3, 0, {d0, d2}, context);
+  AffineMap mapRHS = AffineMap::get(3, 0, {d1, d2}, context);
+  AffineMap mapOut = AffineMap::get(3, 0, {d0, d1}, context);
+  SmallVector<AffineMap> affineMaps{mapLHS, mapRHS, mapOut};
+  return affineMaps;
```
```suggestion
  return {mapLHS, mapRHS, mapOut};
```

kuhar commented on mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:
```diff
@@ -3881,6 +3882,172 @@ Speculation::Speculatability MatmulOp::getSpeculatability() {
+SmallVector<AffineMap>
+BatchMatmulTransposeAOp::getAffineMaps(OpBuilder &builder) {
+  AffineExpr d0, d1, d2, d3;
+  auto context = builder.getContext();
```
```suggestion
  MLIRContext *context = builder.getContext();
```

kuhar reviewed (commented):
Thanks for pushing this forward, just some comments / suggestions for the 'transition' code

rengolin commented:
https://github.com/iree-org/iree/pull/21344 shows the effect on iree. Other than the classof problem @kuhar spotted, seems fine.

rengolin commented:
> So after [this patch](https://github.com/llvm/llvm-project/pull/147168) it'd warrant that `create` methods for the variants be implemented else there'd be a compilation failure for `replaceOpWithNewOp` method.

I've added the create and have used it in the `replaceOpWithNewOp` pattern in `TransposeMatmul.cpp`, which worked fine without the need for the "implicit" version. Is there something else we need to do here?

rengolin commented:
Ok, there were some rebase conflicts that I have resolved, and squashed all into one commit to make it easier to continue to rebase until it gets merged.

rengolin committed:
[MLIR][Linalg] Remove matmul_transpose variants

Removes the `(batch_)matmul_transpose_{a|b}` variants from OpDSL and
replace it with `matmul affine_maps [...]` whenever appropriate.

Added specialization for transpose variants:

No code changes needed, all handled by the new specialization
classes. Fixing a bug in the previous matmul builder where the affine
map was being ignored.

Checking that the affine map is the correct one for the transpose
variant to make sure classof works as expected.

Also exposing a method to check the affine map in case one needs a check
for `matmul && isTransposeA` to replace the old checks for
`matmul_transpose_a` and others.

Adding new create methods following #147168

Common API for all matmul variants

Moved isExpectedAffineMaps -> isDefaultIndexingMaps
Moved getAffineMaps -> getDefaultIndexingMaps
Added isDefaultIndexingMaps to the core ops (ODS)

rengolin commented:
@kuhar @banach-space @qedawkins this should be ready to merge.
